packages/core/src/components/dialog/dialog.tsx

```diff
@@ -121,14 +126,19 @@ export class Dialog extends AbstractPureComponent2<DialogProps> {
             <Overlay {...this.props} className={Classes.OVERLAY_SCROLL_CONTAINER} hasBackdrop={true}>
                 <div className={Classes.DIALOG_CONTAINER} ref={this.props.containerRef}>
                     <div
-                        className={classNames(Classes.DIALOG, this.props.className)}
+                        className={classNames(Classes.DIALOG, this.props.footer != null && Classes.DIALOG_NO_PADDING, this.props.className)}
                         role="dialog"
                         aria-labelledby={this.props["aria-labelledby"] || (this.props.title ? this.titleId : undefined)}
                         aria-describedby={this.props["aria-describedby"]}
                         style={this.props.style}
                     >
                         {this.maybeRenderHeader()}
-                        {this.props.children}
+
+                        <div className={Classes.DIALOG_BODY_SCROLL_CONTAINER}>
+                            {this.props.children}
+                        </div>
```

Comment:
I'm afraid this may be a breaking change for some consumers who are applying their own scroll/overflow styles to dialogs. We will have to add it in a non-breaking way with a new prop, something like `useScrollableBodyContainer?: boolean` where the default is `false` to preserve existing behavior.

Or: see my comment below with an API proposal with 3 new components.